### PR TITLE
HPCC-18591 Fix warnings from gcc 7.2 (which are treated as errors)

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [x] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
@xwang2713 please review

File: esp/bindings/SOAP/xpp/sxt/XmlTokenizer.h [hunk 1]
```diff
@@ -222,7 +222,7 @@ namespace sxt {
    *
    */
    //int XmlTokenizer::next() throw(XmlTokenizerException) {
-   int next() throw(XmlTokenizerException) {
+   int next() {
```
Comment (ghalliday) [inline esp/bindings/SOAP/xpp/sxt/XmlTokenizer.h]: throw specifiers are deprecated in c++11

File: plugins/stringlib/stringlib.cpp [hunk 2]
```diff
@@ -1580,6 +1580,11 @@ STRINGLIB_API unsigned STRINGLIB_CALL slMatchTimeOfDay(size32_t lenS, const char
     return 0;
 }
 
+#ifdef __GNUC__
```
Comment (ghalliday) [inline plugins/stringlib/stringlib.cpp]: I tried using __attribute__((format)) but couldn't get it to work.

File: system/jlib/jptree.cpp [hunk 3]
```diff
@@ -4016,7 +4016,7 @@ class CommonReaderBase : public CInterface
             c++;
         }
     }
-    void error(const char *msg=NULL, bool giveContext=true, PTreeReadExcptCode code=PTreeRead_syntax)
+    [[noreturn]] void error(const char *msg=NULL, bool giveContext=true, PTreeReadExcptCode code=PTreeRead_syntax)
```
Comment (ghalliday) [inline system/jlib/jptree.cpp]: The c++11 way of adding __attribute__((x)).  Tested with VS2015, VS2017 new and older gccs.

**Comment by ghalliday** [comment]
This has an issue in code I rebased onto which I will need to fix tomorrow.

File: plugins/stringlib/stringlib.cpp (hunk 2 above)
Comment (xwang2713) [reply plugins/stringlib/stringlib.cpp]: I use 
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wformat-nonliteral"
 strftime ...
#pragma GCC diagnostic pop

Don't know other way except disable format-literal checking globally.  Yes attribute ((format)) doesn't work for strftime.

File: esp/bindings/SOAP/xpp/sxt/XmlTokenizer.h (hunk 1 above)
Comment (xwang2713) [reply esp/bindings/SOAP/xpp/sxt/XmlTokenizer.h]: I need this when enable "std=c++1z" otherwise just warning.

**Comment by xwang2713** [comment]
Fallthrough checking probably need disabled for plugins/Cassandra/ since we don't want to modify cpp-driver code now. I add it in plugins/Cassandra/CMakeLists.txt.

I have a branch which compiled OK on Ubuntu 17.10.  The fixes include most of my changes.
I am building on other Linux distros. Will try other projects such as plugins, clienttools, EE, LN and do basic testing

**Review by xwang2713 — COMMENTED** [review]
"-Werror=maybe-uninitialized" for plugins/cassandra/CMakeLists.txt otherwise compile fails with:  /var/lib/jenkins/workspace/CE-JIRA-18591-GCC-7-GH/CE/ubuntu-17.10-amd64/HPCC-Platform/plugins/cassandra/cpp-driver/src/buffer.hpp:199:16: error: ‘*((void*)(&<anonymous>)+8).cass::Buffer::data_.cass::Buffer::<unnamed union>::buffer’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
13:39:26      RefBuffer* temp = data_.buffer;

build log for branch issue18591: http://10.240.32.243/view/custom/job/CE-JIRA-18591-GCC-7-GH/CE=ubuntu-17.10-amd64/1/console

**Comment by ghalliday** [comment]
@xwang2713 which version of gcc is that?  Mine works with "gcc (Ubuntu 7.2.0-8ubuntu3) 7.2.0"

**Comment by xwang2713** [comment]
gcc (Ubuntu 7.2.0-8ubuntu3) 7.2.0

**Comment by ghalliday** [comment]
Fixed by https://github.com/hpcc-systems/HPCC-Platform/pull/10767

**Pull request closed without merging**